keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -0,0 +1,337 @@
+cause all connections on that node to get terminated instantly. This can be
+considered a bug / sub-optimal behavior for nodes which are experiencing
+transient readiness state or for terminating nodes, since connections are not
+allowed to drain in those cases, even though the load balancer might support
+that. Moreover: on large clusters with a lot nodes and entropy, re-syncing load
+balancers like this can lead to rate-limiting by the cloud provider due to an
+excessive amount of update calls.
+
+As to enable connection draining, reduce cloud provider API calls and simplify
+the KCCMs sync loop: this KEP proposes that the service controller stops
+synchronizing the load balancer node set in these cases. Seeing as how this has
+always been the case, a new feature gate `LoadBalancerConnectionDraining` will
+be introduced, which will be used to enable the more optimal behavior.
+
+## Motivation
+
+Abruptly terminating connections in the cases defined by a) and b) above can be
+seen as buggy behavior and should be improved. By enabling connection draining,
+applications are allowed profit from graceful shutdown / termination, for what
+concerns cluster ingress connectivity. Users of Kubernetes will also see a
+reduction in the amount of cloud API calls, for what concerns calls stemming
+from syncing load balancers with the Kubernetes cluster state.
+
+### Goals
+
+- Stop re-configuring the load balancers' node set for cases a) and b) above
+
+### Non-Goals
+
+- Stop re-configuring the load balancers' node set for fully deleted /
+  newly added cluster nodes, or for nodes which get annotated with
+  `node.kubernetes.io/exclude-from-external-load-balancers`.
+
+## Proposal
+
+### Risks and Mitigations
+
+#### Risk
+
+- Cloud providers which do not allow VM deletion when the VM is referenced by
+  other constructs, will block the cluster auto-scaler (CA) from deleting the VM
+  upon downscale. This will result in reduced downscale performance by the CA,
+  or completely block VM deletion from happening - this is because the service
+  controller will never proceed to de-reference the VM from the load balancer
+  node set until the Node is fully deleted in the API server, which will never
+  occur until the VM is deleted. The three major cloud providers (GCP/AWS/Azure)
+  do however support this, and it is not expected that other providers don't.
+- By removing the b) predicate above we are delaying the removal of the node
+  from the load balancers' node set until the node is completely deleted in the
+  API server. Workloads relying on this behavior will observe a change in
+  behavior, but this also not expected.
+
+#### Mitigations
+
+- Cloud providers/workloads which do not support the behavior mentioned in
+  (Risk)[#risk], have the possibility to set the feature flag to false, thus
+  default back to the current mechanism.
+
+## Design Details
+
+- Implement the change in the service controller and ensure it does not add /
```

Comment:
In one of the discussions we suggested not *adding* a node if it was unready or CA-tainted, but once it was added, not removing it for those conditions.  It seems unlikely to be an actual issue, so the complexity may not be justified, but I think we should call it out.

Specifically, if a new Node comes online, but it is unready - do we add it to the LB set or wait for it to be ready?

Reply:
> Specifically, if a new Node comes online, but it is unready - do we add it to the LB set or wait for it to be ready?

I am thinking, if we:

* Add node on ADD (and `NotReady`): then it's in the set but LB will follow HC/service proxy signal 
   * ++ faster reaction time for correct traffic load balancing.   
   * ++ less complexity in service controller   
* Add node on UPDATE (and `Ready`): then it's (probably) not in the set until after the service proxy is reporting green  
   * -- slower reaction time for correct traffic load balancing.   
   * -- more complexity in service controller 

The only thing which might speak against the ADD case is: if the CNI plugin fails completely (i.e: unable to write it's config to disk) causing the node to remain `NotReady` indefinitely; the service proxy is unaware of this and might report green to the LB. But in that case we are sure we won't have the endpoint running on the Node in question, so eTP:Local is fine and for eTP:Cluster: well, the service proxy is fine so it can forward packets independently of the CNI plugin...unless the CNI plugin has completely screwed up the host's network while crashing, impacting kube-proxy? But in that case the LB won't probably even be able to reach kube-proxy / service-proxy.

Reply:
See above https://github.com/kubernetes/enhancements/pull/3460/files#r1095135706 for more on why we might want to do (a form of) this